style: edtlib: Use a more efficient expression

rruuaanng opened the pull request:
Use a list comprehension to replace nested
loops for better execution speed.

mbolivar reviewed (changes requested):
Since this PR is claiming a performance optimization, I would like to see some benchmarks proving it's actually an improvement.

Also, why is the `from pathlib import Path` down in the function instead of at top of file? That seems weird to me.

rruuaanng commented:
Actually, Python internally optimizes the append semantics for list comprehensions, making them superior to regular for loops in any scenario. For the this benchmark:
```
ruang@forNova:~$ cat ben.py 
import time

n = 1000000

def normal_for_loop():
    result = []
    for i in range(n):
        result.append(i * 2)
    return result

def list_comprehension():
    return [i * 2 for i in range(n)]

start_time = time.time()
normal_for_loop()
end_time = time.time()
print(f"normal for loop: {end_time - start_time:.6f}s")

start_time = time.time()
list_comprehension()
end_time = time.time()
print(f"comprehension: {end_time - start_time:.6f}s")

ruang@forNova:~$ python3 ben.py 
normal for loop: 0.078915s
comprehension: 0.069242s

```

mbolivar commented:
> Python internally optimizes the append semantics for list comprehensions, making them superior to regular for loops in any scenario. For the this benchmark:

Thanks! Can you include a similar benchmark in the commit log that applies to our code here? It should be easy to hack a patch right before and after this assignment in edtlib.py:

```py
        self._binding_paths: list[str] = _binding_paths(self.bindings_dirs)
```

> This is a scoped import that exists only in the current function and not in the global symbol table.

I understand the syntax, but what I'm saying is that we don't tend to use these in zephyr's python code, and I'm not sure why this commit should be making an exception to the conventional style. This feels like the sort of thing that doesn't scale well when copy/pasted, so we should have a compelling reason (like "this import takes a long time and is only needed in uncommon cases, so this is a performance improvement") to add it. IMO

rruuaanng commented:
> > Python internally optimizes the append semantics for list comprehensions, making them superior to regular for loops in any scenario. For the this benchmark:
> 
> Thanks! Can you include a similar benchmark in the commit log that applies to our code here? It should be easy to hack a patch right before and after this assignment in edtlib.py:
> 
> ```python
>         self._binding_paths: list[str] = _binding_paths(self.bindings_dirs)
> ```
> 
> > This is a scoped import that exists only in the current function and not in the global symbol table.
> 
> I understand the syntax, but what I'm saying is that we don't tend to use these in zephyr's python code, and I'm not sure why this commit should be making an exception to the conventional style. This feels like the sort of thing that doesn't scale well when copy/pasted, so we should have a compelling reason (like "this import takes a long time and is only needed in uncommon cases, so this is a performance improvement") to add it. IMO


> I understand the syntax, but what I'm saying is that we don't tend to use these in zephyr's python code, and I'm not sure why this commit should be making an exception to the conventional style. This feels like the sort of thing that doesn't scale well when copy/pasted, so we should have a compelling reason (like "this import takes a long time and is only needed in uncommon cases, so this is a performance improvement") to add it. IMO

A reasonable suggestion. I agree, I will move it to the top of the file to maintain consistency with the existing code.

rruuaanng committed:
style: edtlib: Use a more efficient expression

Replaced nested loops with a list comprehension
to improve performance. Execution time improved
from 0.0046203136444s to 0.0040774345397s

Signed-off-by: James Roy <rruuaanng@outlook.com>

mbolivar reviewed (approved):
Looks fine, though I'm guessing that it will take a while for the 0.0005 s  performance improvement to add up to appreciable wins in CI land, every little bit helps. Thanks.
